addons/account/models/account_reconcile_model.py

```diff
@@ -355,267 +367,372 @@ def _prepare_reconciliation(self, st_line, aml_ids=[], partner=None):
-                LEFT JOIN res_company company           ON company.id = st_line_move.company_id
-                LEFT JOIN partners_table line_partner   ON line_partner.line_id = st_line.id
-                WHERE st_line.id IN %s
-            '''
-            params = [rule.sequence, rule.id, tuple(st_lines.ids)]
+        self.ensure_one()
 
-            query, params = rule._apply_conditions(query, params)
-            queries.append(query)
-            all_params += params
+        if self.rule_type != 'writeoff_suggestion':
+            raise UserError(_("Programmation Error: Can't call _get_writeoff_suggestion_query() for different rules than 'writeoff_suggestion'"))
 
-        full_query = self._get_with_tables(st_lines, partner_map=partner_map)
-        full_query += ' UNION ALL '.join(queries)
-        return full_query, all_params
+        query = '''
+            SELECT
+                %(sequence)s                        AS sequence,
+                %(model_id)s                        AS model_id,
+                st_line.id                          AS id
+            FROM account_bank_statement_line st_line
+            JOIN account_move st_line_move          ON st_line_move.id = st_line.move_id
+            LEFT JOIN account_journal journal       ON journal.id = st_line_move.journal_id
+            LEFT JOIN res_company company           ON company.id = st_line_move.company_id
+            WHERE st_line.id IN %(st_line_ids)s
+        '''
+        params = {
+            'sequence': self.sequence,
+            'model_id': self.id,
+            'st_line_ids': tuple(st_lines.ids),
+        }
+
+        return query, params
+
+    def _get_rule_result(self, st_line, candidates, aml_ids_to_exclude, reconciled_amls_ids, partner_map):
+        """ Get the result of a rule from the list of available candidates, depending on the
+        other reconciliations performed by previous rules.
+        """
+        self.ensure_one()
+
+        if self.rule_type == 'invoice_matching':
+            return self._get_invoice_matching_rule_result(st_line, candidates, aml_ids_to_exclude, reconciled_amls_ids, partner_map)
+        elif self.rule_type == 'writeoff_suggestion':
+            return self._get_writeoff_suggestion_rule_result(st_line, partner_map), set(), set()
+        else:
+            return None, set(), set()
+
+    def _get_invoice_matching_rule_result(self, st_line, candidates, aml_ids_to_exclude, reconciled_amls_ids, partner_map):
+        new_reconciled_aml_ids = set()
+        new_treated_aml_ids = set()
+        candidates, priorities = self._filter_candidates(candidates, aml_ids_to_exclude, reconciled_amls_ids)
+
+        # Special case: the amounts are the same, submit the line directly.
+        st_line_currency = st_line.currency_id or st_line.journal_id.currency_id or st_line.company_id.currency_id
+        st_line_residual = st_line.currency_id and st_line.amount_currency or st_line.amount
```

Comment:
```
liquidity_lines, suspense_lines, other_lines = st_line._seek_for_lines()
st_line_currency = suspense_lines.currency_id or suspense_lines.company_currency_id
balance_field = 'amount_currency' if suspense_lines.currency_id else 'balance'
residual_field =  'amount_residual_currency' if suspense_lines.currency_id else 'amount_residual'

if st_line.to_check:
   st_line_residual = -liquidity_lines[balance_field]
elif suspense_lines.account_id.reconcile:
   st_line_residual = sum(suspense_lines[residual_field])
else:
   st_line_residual = sum(suspense_lines[balance_field])
```

Note: only one suspense line is allowed
Note2: It could be added in statement line directly since getting the residual amount is quite complicated

Reply:
Done here, as it's more a fix for the blue lines: https://github.com/odoo/odoo/pull/51153/commits I'm already integrating it to this branch as well.